### ISSUE TITLE
[tune] [wandb] Experiment checkpointing fails with `WandbTrainableMixin`

### ISSUE THREAD

**Issue by konichuvak** [ray-project/ray#11917]
### What is the problem?
Experiment checkpointing fails when `WandbTrainableMixin` is added to the `Trainable` base classes.
The [signature](https://github.com/ray-project/ray/blob/releases/1.0.0/python/ray/tune/integration/wandb.py#L392) of `WandbTrainableMixin.__init__` is `config, *args, **kwargs` which makes it look like a checkpointable function (i.e. it passes `func(config, checkpoint_dir=None)`) when validated by `detect_checkpoint_function` in [`Experiment.__init__`](https://github.com/ray-project/ray/blob/releases/1.0.0/python/ray/tune/experiment.py#L127). 

### Reproduction
```python
import numpy as np
import wandb
from ray import tune
from ray.tune import Trainable
from ray.tune.integration.wandb import WandbTrainableMixin


class WandbTrainable(WandbTrainableMixin, Trainable):
    def step(self):
        for i in range(30):
            loss = self.config["mean"] + self.config["sd"] * np.random.randn()
            wandb.log({"loss": loss})
        return {"loss": loss, "done": True}


if __name__ == "__main__":
    tune.run(
        WandbTrainable,
        checkpoint_freq=1,
        config={
            "mean": tune.grid_search([1, 2, 3, 4, 5]),
            "sd": tune.uniform(0.2, 0.8),
            "wandb": {
                "api_key": 'WANDB_API_KEY',
                "project": "example-project"
            }
        })
```

#### Suggested solutions

- The quickest workaround is to explicitly re-define a signature in the user's `Trainable` class, such that it is no longer identified as a checkpointable function:
```
    def __init__(self, config=None, logger_creator=None):
        super().__init__(config=config, logger_creator=logger_creator)
```

- Change the signature of `WandbTrainableMixin` to mirror that of `Trainable`:
```
    def __init__(self, config: Dict, logger_creator=None):
```

---

If we cannot run your script, we cannot fix your issue.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

**Comment by richardliaw** [comment]
Hmm, I think rather we should just detect that the input is a Trainable class and return False in `detect_checkpoint_function`. Would that make sense?

cc @krfricke

**Comment by konichuvak** [comment]
That works! We can add the following check to `detect_checkpoint_function`:
```python
    if isinstance(train_func, type) and issubclass(train_func, ray.tune.Trainable):
        return False
```

However, if the definition of checkpointable function always requires just 2 arguments (config, checkpoint_dir), then we could replace the try/except block with a direct check.

That is, we can use 
```python
    if set(func_sig.parameters) - {'self'} != {'config', 'checkpoint_dir'}:
        logger.debug(f"Invalid signature for checkpoint function {func_sig}")
```
instead of 
```python
    try:
        # check if signature is func(config, checkpoint_dir=None)
        func_sig.bind({}, checkpoint_dir="tmp/path")
    except Exception as e:
        logger.debug(str(e))
```
I could take a stab at PR in case either of the two looks reasonable.

**Comment by richardliaw** [comment]
Hmm, yeah I think option 1 is more robust, so let’s do that. Ping me if you
need help with anything!

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
